packages/connector/src/services/core/controllers/stream.ts

```diff
@@ -33,15 +33,18 @@ export function createStreamController(): RafikiMiddleware {
 
     const { connectionId } = moneyOrReply
     const connectionKey = streamReceivedKey(connectionId)
-    // Pass the amount as a string so that values higher than MAX_SAFE_INTEGER aren't modified by a local float conversion.
-    // Ignore the running total returned by `incrby`, since it is parsed to a `Number`, which may
-    // be inaccurate if the value is higher than MAX_SAFE_INTEGER.
-    await redis.incrby(
-      connectionKey,
-      (request.prepare.amount as unknown) as number
-    )
-    await redis.expire(connectionKey, CONNECTION_EXPIRY)
-    const totalReceived = await redis.get(connectionKey)
+    // Thanks to Redis's `stringNumbers:true`, `incrby` returns a string rather than a number.
+    // This ensures that precision isn't lost when dealing with integers larger than MAX_SAFE_INTEGER.
+    const [[err, totalReceived], [err2]] = await redis
+      .multi()
+      .incrby(
+        connectionKey,
+        (request.prepare.amount.toString() as unknown) as number
+      )
+      .expire(connectionKey, CONNECTION_EXPIRY)
+      .exec()
+    if (err) throw err
```

Comment:
I don't like throwing unexpected, untyped errors, especially for a non-critical function like signing STREAM receipts. (I'm going to open an issue about replacing errors with returning Rejects elsewhere, too. It seems like a source of bugs: if a downstream middleware throws, do upstream ones undo their side effects, such as balance updates, timers, etc.? What if those middlewares need to do things after downstream errors?)

I think these are typically syntax/invalid argument errors -- which shouldn't occur here. One _possible_ case is an overflow error if the connection key hits the i64 max, but that either represents a misconfiguration or something nefarious.

I'd propose either: don't call `setTotalReceived` and still fulfill the packet, or gracefully close the STREAM connection using one of the decline functions. (So, ignore the Redis errors are maybe just confirm `totalReceived` is an integer/non-null?)

Reply:
:+1: I like `don't call setTotalReceived and still fulfill the packet`, combined with logging the error.

That said, in an `async` function isn't `throw foo` effectively `return Promise.reject(foo)` anyway? Then the rejection get turned back into an exception by the caller's `await` -- and so on, until eventually the error-handler middleware turns it into an error response. Not a great system, but I don't see how returning rejects is different.

Reply:
> That said, in an async function isn't throw foo effectively return Promise.reject(foo) anyway? Then the rejection get turned back into an exception by the caller's await -- and so on, until eventually the error-handler middleware turns it into an error response. Not a great system, but I don't see how returning rejects is different.

Yes, I think this is undesirable.

I'm suggesting Fulfills & Rejects would be passed as return values, rather than via mutable context: so middlewares and `next` calls return `Promise<IlpFulfill | IlpReject>`. This enforces that middlewares must call `next` and/or return, and if a middleware has logic dealing with an ILP reply, it handles all Reject cases. While middlewares could currently attach a `catch` handler when they call `next`, handling Promise rejections isn't enforced by TS, and there's no type checking on the rejected values.

Additionally:
- The middleware encountering the error has the most context for which ILP Reject should be returned, so it's preferable for it to decide over a generic error
- Throwing custom ILP errors adds indirection & generates a call stack, whereas returning a specific Reject directly  doesn't have that overhead
- Some middlewares might want to map one Reject to another Reject, or apply side effects on a particular type of Reject

Reply:
Ah, I see what you mean. I was conflating Promise rejects with ILP rejects.
I agree, that's a much better plan. It also makes the real unexpected thrown errors stand out much better. :+1: